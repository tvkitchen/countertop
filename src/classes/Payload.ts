import { isPayloadParameters } from '../types'
import type {
	PayloadParameters,
	PayloadType,
} from '../types'

interface JsonDeserializedBuffer {
	type: string;
	data: string;
}
const isJsonDeserializedBuffer = (value: unknown): value is JsonDeserializedBuffer => (
	value instanceof Object
	&& 'type' in value
	&& 'data' in value
)

export class Payload {
	public readonly data: Buffer

	public readonly type: PayloadType | string

	public readonly createdAt: string

	public readonly origin: string

	public readonly duration: number

	public readonly position: number

	public constructor(parameters: PayloadParameters) {
		this.data = parameters.data
		this.type = parameters.type
		this.createdAt = parameters.createdAt ?? (new Date()).toISOString()
		this.origin = parameters.origin
		this.duration = parameters.duration
		this.position = parameters.position
	}

	public static serialize(payload: Payload): string {
		return JSON.stringify(payload)
	}

	public static deserialize(serializedPayload: string): Payload {
		const deserializedPayload: unknown = JSON.parse(serializedPayload, (_, value: unknown) => {
			if (isJsonDeserializedBuffer(value)
				&& value.type === 'Buffer') {
				return Buffer.from(value.data)
			}
			return value
		})
		if (isPayloadParameters(deserializedPayload)) {
			return new Payload(deserializedPayload)
		}
		throw new Error(`Invalid payload serialiation: ${serializedPayload}`)
	}
}
