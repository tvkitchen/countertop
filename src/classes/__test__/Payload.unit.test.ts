import { Payload } from '../Payload'

describe('Payload', () => {
	describe('constructor', () => {
		it('should construct when provided all parameters', () => {
			const parameters = {
				data: Buffer.from('I ate all of the cheese'),
				type: 'CONFESSION',
				createdAt: '2020-02-02T03:04:05.000Z',
				origin: '2020-02-02T03:04:01.000Z',
				duration: 1000,
				position: 60000,
			}
			const payload = new Payload(parameters)
			expect(payload).not.toBeUndefined()
			expect(payload.data.toString()).toBe('I ate all of the cheese')
			expect(payload.type).toBe('CONFESSION')
			expect(payload.createdAt).toEqual('2020-02-02T03:04:05.000Z')
			expect(payload.origin).toEqual('2020-02-02T03:04:01.000Z')
			expect(payload.duration).toBe(1000)
			expect(payload.position).toBe(60000)
		})
		it('should populate a default createdAt', () => {
			const parameters = {
				data: Buffer.from('I ate all of the cheese'),
				type: 'CONFESSION',
				origin: '2020-02-02T03:04:01.000Z',
				duration: 1000,
				position: 60000,
			}
			const payload = new Payload(parameters)
			expect(payload.createdAt).not.toBe('')
		})
		it('should set createdAt differently for payloads created at different times', () => {
			jest.clearAllMocks()
			const parameters = {
				data: Buffer.from('I ate all of the cheese'),
				type: 'CONFESSION',
				origin: '2020-02-02T03:04:01.000Z',
				duration: 1000,
				position: 60000,
			}
			jest.setSystemTime(new Date('1592156234000'))
			const firstPayload = new Payload(parameters)
			jest.setSystemTime(new Date('1592156235000'))
			const secondPayload = new Payload(parameters)
			expect(firstPayload.createdAt).not.toEqual(secondPayload.createdAt)
		})
	})

	describe('serialize', () => {
		it('should properly should serialize data with a fully populated object', () => {
			const parameters = {
				data: Buffer.from('I ate all of the cheese'),
				type: 'CONFESSION',
				createdAt: '2020-02-02T03:04:05.000Z',
				origin: '2020-02-02T03:04:01.000Z',
				duration: 1000,
				position: 60000,
			}
			const payload = new Payload(parameters)
			const serializedPayload = Payload.serialize(payload)
			expect(serializedPayload).toMatchSnapshot()
		})
	})

	describe('deserialize', () => {
		it('should error when deserializing an empty serialized object', () => {
			expect(() => Payload.deserialize('{}')).toThrow()
		})
		it('should error when deserializeing a partially populated serialized payload', () => {
			expect(() => Payload.deserialize('{"type":"WHOOPS"}')).toThrow()
		})
		it('should properly deserialize a serialized payload', () => {
			const parameters = {
				data: Buffer.from('I ate all of the cheese'),
				type: 'CONFESSION',
				createdAt: '2020-02-02T03:04:05.000Z',
				origin: '2020-02-02T03:04:01.000Z',
				duration: 1000,
				position: 60000,
			}
			const payload = new Payload(parameters)
			const serializedPayload = Payload.serialize(payload)
			const deserializedPayload = Payload.deserialize(serializedPayload)
			expect(deserializedPayload).toEqual(payload)
		})
		it('should error when deserializing serialized objects with invalid values', () => {
			const parameters = {
				data: 'this is not a buffer',
				type: 'CONFESSION',
				createdAt: '2020-02-02T03:04:05.000Z',
				origin: '2020-02-02T03:04:01.000Z',
				duration: 1000,
				position: 60000,
			}
			const serializedValues = JSON.stringify(parameters)
			expect(() => Payload.deserialize(serializedValues)).toThrow()
		})
	})
})
