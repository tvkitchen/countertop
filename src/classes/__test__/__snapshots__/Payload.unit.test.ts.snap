// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payload serialize should properly should serialize data with a fully populated object 1`] = `"{\\"data\\":{\\"type\\":\\"Buffer\\",\\"data\\":[73,32,97,116,101,32,97,108,108,32,111,102,32,116,104,101,32,99,104,101,101,115,101]},\\"type\\":\\"CONFESSION\\",\\"createdAt\\":\\"2020-02-02T03:04:05.000Z\\",\\"origin\\":\\"2020-02-02T03:04:01.000Z\\",\\"duration\\":1000,\\"position\\":60000}"`;
